pkg/query/columnquery.go

```diff
@@ -861,14 +922,140 @@ func ComputeDiff(ctx context.Context, tracer trace.Tracer, base, compare profile
+					profileLabels = append(profileLabels, field)
+					profileLabelColumns = append(profileLabelColumns, r.Column(i))
+				}
+			}
+
+			cols := make([]arrow.Array, len(profileLabels)+3) // locations, value, diff
+			copy(cols, profileLabelColumns)
+
+			cols[len(cols)-3] = r.Column(schema.FieldIndices("locations")[0])
+			cols[len(cols)-2] = normalizeForDiff(mem, r, totalPeriod, 1)
+
+			// This is intentional, the diff value of the `compare` profile is the same
+			// as the value of the `compare` profile, because what we're actually doing
+			// is subtracting the `base` profile, but the actual calculation happens
+			// when building the visualizations. We should eventually have this be done
+			// directly by the query engine.
+			cols[len(cols)-1] = cols[len(cols)-2]
+
+			records = append(records, array.NewRecord(
+				profile.ArrowSchema(profileLabels),
+				cols,
+				r.NumRows(),
+			))
+		}()
+	}
+
+	for _, r := range base.Samples {
+		func() {
+			schema := r.Schema()
+			profileLabels := []arrow.Field{}
+			profileLabelColumns := []arrow.Array{}
+			for i, field := range schema.Fields() {
+				if strings.HasPrefix(field.Name, profile.ColumnPprofLabelsPrefix) {
+					profileLabels = append(profileLabels, field)
+					profileLabelColumns = append(profileLabelColumns, r.Column(i))
+				}
+			}
+
+			cols := make([]arrow.Array, len(profileLabels)+3) // locations, value, diff
+			copy(cols, profileLabelColumns)
+
+			cols[len(cols)-3] = r.Column(schema.FieldIndices("locations")[0])
+			cols[len(cols)-2] = zeroArray(mem, int(r.NumRows()))
+			cols[len(cols)-1] = normalizeForDiff(mem, r, totalPeriod, -1)
+
+			records = append(records, array.NewRecord(
+				profile.ArrowSchema(profileLabels),
+				cols,
+				r.NumRows(),
+			))
+		}()
+	}
+
+	// We scale the smaller profile up to the bigger one. Therefore, we should use the bigger duration and period.
+	compare.Meta.Period = max(base.Meta.Period, compare.Meta.Period)
+	compare.Meta.Duration = max(base.Meta.Duration, compare.Meta.Duration)
+	compare.Meta.Timestamp = min(base.Meta.Timestamp, compare.Meta.Timestamp)
+
+	return profile.Profile{
+		Meta:    compare.Meta,
+		Samples: records,
+	}, nil
+}
+
+func normalizeForDiff(mem memory.Allocator, r arrow.Record, totalPeriod, factor int64) *array.Int64 {
+	b := array.NewInt64Builder(mem)
+	defer b.Release()
+
+	valueSum := r.Column(r.Schema().FieldIndices(profile.ColumnValueSum)[0]).(*array.Int64)
+	valueCount := r.Column(r.Schema().FieldIndices(profile.ColumnValueCount)[0]).(*array.Int64)
+	period := r.Column(r.Schema().FieldIndices(profile.ColumnPeriod)[0]).(*array.Int64)
+
+	for i := 0; i < valueSum.Len(); i++ {
+		if valueSum.IsValid(i) && period.IsValid(i) {
+			vs := valueSum.Value(i)
+			vc := valueCount.Value(i)
+			p := period.Value(i)
+
+			normValue := factor * (vs / vc)
+			normPeriod := float64(totalPeriod) / float64(p)
+
+			v := float64(normValue) * normPeriod
+			b.Append(int64(v))
```

Comment:
This is the most important part we should check and discuss. 
I'll add some more tests before merging, which should help too.